### ISSUE TITLE
driverless doesn't generate a ppd for dns-sd based uri

### ISSUE THREAD

**Issue by zdohnal** [OpenPrinting/cups-filters#308]
Hi,

we have a report in [Fedora](https://bugzilla.redhat.com/show_bug.cgi?id=1879147) - dns-sd based uris don't work anymore with driverless:

```
$ driverless ipp://HP%20LaserJet%20M1536dnf%20MFP%20%2842307C%29._ipp._tcp.local/

ERROR: Unable to create PPD file: Could not poll sufficient capability info from the printer (ipp://HP%20LaserJet%20M1536dnf%20MFP%20%2842307C%29._ipp._tcp.local/, ipp://NPI42307C.local:631/ipp/printer) via IPP!
```

It seems there are two issues:
1) ```while``` loop doesn't work for empty lines
2) ```buffer``` is uninitialized 

This error blocks adding a print queue via driverless backend in Gnome Control Center and CUPS web ui (when you use dns-sd based uri).

The pull request will come in a minute.

**Comment by zdohnal** [comment]
Fixed via #309 .

**Comment by davex25** [comment]
Although driverless --version reports 1.28.10, and cups-filters package in arch is cups-filters-1.28.10-2, behavior is still happening on my Archlinux system. Here's what I'm seeing:

driverless ipps://EPSON%20ET-2760%20Series._ipps._tcp.local/
ERROR: Unable to create PPD file: Could not poll sufficient capability info from the printer (ipps://EPSON%20ET-2760%20Series._ipps._tcp.local/, ipps://EPSON4EC4A4.local:631/ipp/print) via IPP!

Strangely on my laptop also running Arch, with the same versions as given above, the driverless command works as expected. It seems like removing cups and cups-filters, and reinstalling with latest versions should have fixed this but no luck. Where else could this problem be coming from?

**Comment by tillkamppeter** [comment]
@davex25 this is not the original bug reported here as I have checked now and the fix is not only applied to the master branch but also to 1.x, so 1.28.10 should contain the fix.

**Comment by tillkamppeter** [comment]
I am also not able to reproduce it on Ubuntu.

Does `ipptool -tv ipps://EPSON4EC4A4.local:631/ipp/print get-printer-attributes.test > attrs.txt` work for you and produces something useful in `attrs.txt`?

Please attach your `attrs.txt` file.

**Comment by davex25** [comment]
ipptool: Unable to connect to "EPSON4EC4A4.local" on port 631 - Host is down

Tried the same command from my laptop on the LAN, and got this: 
[attrs.txt](https://github.com/OpenPrinting/cups-filters/files/7878007/attrs.txt)

**Comment by tillkamppeter** [comment]
@davex25 thank you. The output looks correct, but it seems that your printer can only be accessed from your laptop and not from your desktop. As you say it is probably an issue of your network and not a bug in cups-filters.

**Comment by davex25** [comment]
In case someone else runs into this, I discovered that the problem was Avahi was misconfigured on my system. Package "nss-mdns-0.15.1-1" was not installed, and hosts line in nsswitch.conf was not using mdns, mdns4, or mdns4_minimal. 

Everything works after following instructions from the hostname resolution section of https://wiki.archlinux.org/title/avahi.

**Comment by RickOrchard** [comment]
> In case someone else runs into this, I discovered that the problem was Avahi was misconfigured on my system. Package "nss-mdns-0.15.1-1" was not installed, and hosts line in nsswitch.conf was not using mdns, mdns4, or mdns4_minimal.
> 
> Everything works after following instructions from the hostname resolution section of https://wiki.archlinux.org/title/avahi.

I ran into the same issue on my Endeavour install. Indeed, installing Avahi and nss-mdns package (dont forget to start the service) and manually editing the nsswitch.conf file did the trick! Man was i searching for this answer, thanks.

Never would have thought that the 'cups-internal-server-error' was related to a missing DNS service...

**Comment by skerit** [comment]
Same issue here on Endeavouros. Did all the steps and now it's working.

**Comment by makitso** [comment]
Ubuntu Budgie has this same issue with 22.04.2, 22.10 and 23.04

**Comment by BvngeeCord** [comment]
> In case someone else runs into this, I discovered that the problem was Avahi was misconfigured on my system. Package "nss-mdns-0.15.1-1" was not installed, and hosts line in nsswitch.conf was not using mdns, mdns4, or mdns4_minimal.
> 
> Everything works after following instructions from the hostname resolution section of [wiki.archlinux.org/title/avahi](https://wiki.archlinux.org/title/avahi).

Thank you so much. This was exactly what my problem was. You saved me from hours of suffering

**Comment by zdohnal** [comment]
It would be great if the affected distros added nss-mdns+avahi at least as a weak dependency. mDNS is an easy way for common users to get their printers working.

**Comment by luyatshimbalanga** [comment]
> I am getting this error today. Probably since one of last updates of packages.
> 
> I use Arch. The printer worked until recently, no issues. Even tested occasionally by removing it and adding it back.
> 
> I think it happened since the update with the recent CUPS security issue.
> 

Hello, I got similar issue after updating to the latest CUPS drive on Fedora 41 with a HP OfficeJet Pro printer.
The solution is to update `authselect` profile with the following command:
`sudo authselect select local with-silent-lastlog with-mdns4`

That will update /etc/nsswitch.conf with the following profile
```
# Generated by authselect
# Do not modify this file manually, use authselect instead. Any user changes will be overwritten.
# You can stop authselect from managing your configuration by calling 'authselect opt-out'.
# See authselect(8) for more details.

# In order of likelihood of use to accelerate lookup.
passwd:     files systemd
shadow:     files
group:      files [SUCCESS=merge] systemd
hosts:      files myhostname mdns4_minimal [NOTFOUND=return] resolve [!UNAVAIL=return] dns
services:   files
netgroup:   files
automount:  files

aliases:    files
ethers:     files
gshadow:    files
networks:   files dns
protocols:  files
publickey:  files
rpc:        files
```

**Comment by fuxoft** [comment]
Hello, I have installed XUbuntu 24.04 and I am experiencing exactly this bug, unable to add my HP 9022 network printer which worked flawlessly with previous XUbuntu versions (and still works from other PCs in my household).

Unfortunately, I am unable to apply the nsselect solution listed here:

```
$ lpadmin -p hp9020 -E -v dnssd://BonjourHPOfficeJetPro9022e._ipp._tcp.local/?uuid=72692038-5a09-4086-9ab9-369ab05254f8 -m everywhere
lpadmin: Unable to connect to HP14CB1936C7AC.local:631: Host is down

$ sudo authselect select local with-silent-lastlog with-mdns4
[error] /etc/authselect/authselect.conf is missing or unreadable, system was not properly configured by authselect.
[error] Refusing to activate profile unless overwrite is requested.

```

My /etc/nsswitch.conf looks like this:

```
# /etc/nsswitch.conf
#
# Example configuration of GNU Name Service Switch functionality.
# If you have the `glibc-doc-reference' and `info' packages installed, try:
# `info libc "Name Service Switch"' for information about this file.

passwd:         files systemd
group:          files systemd
shadow:         files systemd
gshadow:        files systemd

hosts:          files mdns4_minimal [NOTFOUND=return] dns
networks:       files

protocols:      db files
services:       db files
ethers:         db files
rpc:            db files

netgroup:       nis

```

Unfortunately I have no idea what "nsselect" or "avahi" is. :(

**Comment by zdohnal** [comment]
@fuxoft Hello Mr. Fuka,

it looks like mDNS resolution does not work as it should - you can check it by pinging that .local mDNS hostname:

```
$ ping HP14CB1936C7AC.local
```

Your device does not use authselect for handling `/etc/nsswitch.conf`, but in general the file looks ok (though I'm not a mDNS/NSS expert...), so IMHO the problem will be on the side of Avahi/firewall.

Usually, avahi-daemon systemd service handles running Avahi (so if you use systemd, make sure the socket unit is running - `sudo systemctl enable --now avahi-daemon.socket`) and the machine has to have mDNS port opened in firewall if there is any (check `firewalld` service if it is running, and if so, check if port 5353 is opened), but I'm not sure what are XUbuntu default setup/settings, so you might have to search a little how it is for XUbuntu.

**Comment by fuxoft** [comment]
Thank you, @zdohnal 

`ping HP14CB1936C7AC.local` indeed does not work.

There does not seem to be any firewall on my machine, at least I don't see any in `sudo service --status-all` and `sudo ufw status` says `Status: inactive`.

I have entered the command you suggested: `sudo systemctl enable --now avahi-daemon.socket` - It produces no output and PINGing my printer (as above) still does not work: The result is `ping: HP14CB196C7AC.local: Temporary failure in name resolution`.

**Comment by fuxoft** [comment]
SOLVED! Thanks, @zdohnal !

The problem was, that - by default - MDNS on Xubuntu is disabled!

So I've added the following to `/etc/systemd/resolved.conf`:

```
[Resolve]
...
MulticastDNS=yes
```

And then I did: `sudo resolvectl mdns enp3s0 yes`

Restarted, and the printer works.

That seems like serious oversight from the XUbuntu team...

Until yesterday, I had no idea what "mdns" is so anyone feel free to correct me.

**Comment by zdohnal** [comment]
@fuxoft in general you wouldn't have to enable mDNS in systemd-resolved, it should work with avahi-daemon+nss-mdns as it is, but that's more question for Ubuntu folk why it does not work.
If you go via resolved, I would recommend enabling mDNS only on the network interface you want to use mDNS on rather than in global settings - resolved provides this granularity of choice, once mDNS is enabled on the specific network interface in NetworkManager too.

**Comment by fuxoft** [comment]
> [@fuxoft](https://github.com/fuxoft) in general you wouldn't have to enable mDNS in systemd-resolved, it should work with avahi-daemon+nss-mdns as it is, but that's more question for Ubuntu folk why it does not work.

@zdohnal , at first I tried just `sudo resolvectl mdns enp3s0 yes` but it complained that "mdns is set to globally OFF" so I did that thing with resolved.conf that I googled elsewehere. Later, I've done `sudo resolvectl mdns ztuga3w3o2 off` to switch mdns off for my Zerotier one interface. I hope.

I have a bit of understanding about how DNS and DHCP work but, unfortunately, I have zero idea about what Avahi does, how do I configure it and how do I even see if it's active in my system. If someone is willing to contact me privately and help me to configure my system correctly, I am more than willing to accept that offer.
